src/document_parser_callbacks.h

```diff
@@ -89,20 +89,45 @@ really_inline bool parser::on_null_atom() noexcept {
 really_inline uint8_t *parser::on_start_string() noexcept {
   /* we advance the point, accounting for the fact that we have a NULL
     * termination         */
+  // If we limit JSON documents to strictly less 4GB of
+  // string content, then current_string_buf_loc
+  // - doc.string_buf.get() fits in 32 bits. This leaves us
+  // three free bytes.
   write_tape(current_string_buf_loc - doc.string_buf.get(), internal::tape_type::STRING);
-  return current_string_buf_loc + sizeof(uint32_t);
+  return current_string_buf_loc + sizeof(uint16_t);
 }
 
 really_inline bool parser::on_end_string(uint8_t *dst) noexcept {
   uint32_t str_length = uint32_t(dst - (current_string_buf_loc + sizeof(uint32_t)));
   // TODO check for overflow in case someone has a crazy string (>=4GB?)
   // But only add the overflow check when the document itself exceeds 4GB
   // Currently unneeded because we refuse to parse docs larger or equal to 4GB.
-  memcpy(current_string_buf_loc, &str_length, sizeof(uint32_t));
+
+  // We have two scenarios here. Either the string length is
+  // less than 0x7fffff in which case, we have room in the string
+  // header and all is good. Otherwise, we can encode the
+  // string length in the document itself, taking care to
+  // ensure that we do so in ASCII.
+  if(likely(str_length <= 0x7fffff)) { // likely
+    doc.tape[current_loc-1] |=  uint64_t(str_length) << 32;
+    // we have a string header that must be ASCII, unused in
+    // this common case
+    current_string_buf_loc[0] = uint8_t(32); // space
+    current_string_buf_loc[1] = uint8_t(32); // space
+    // we are done!
+  } else {
+    // oh gosh, we have a long string.
+    doc.tape[current_loc-1] |=  uint64_t(0x800000 | (str_length >> 9)) << 32;
+    // we have 9 bits left to code, which we do on the string buffer
+    // using two bytes
+    current_string_buf_loc[0] = uint8_t(32 + ((str_length & 0x1f0) >> 4))
+    current_string_buf_loc[1] = 32 + uint8_t(str_length & 0xf);
+  }
   // NULL termination is still handy if you expect all your strings to
-  // be NULL terminated? It comes at a small cost
-  *dst = 0;
-  current_string_buf_loc = dst + 1;
+  // be NULL terminated? It comes at a small cost and if it is
+  // never used, we might as well drop it.
+  //*dst = 0;
```

Comment:
Is part of the proposal to drop this? I think plenty of people might still want it.

Reply:
... actually, as long as the next string is less than 0x100000000 bytes, if we don't add 32 to the "overflow length," all strings will be null terminated *by default*. And that condition is both unusual and easy to check for :)

Reply:
> actually, as long as the next string is less than 0x100000000 bytes, if we don't add 32 to the "overflow length," all strings will be null terminated by default. And that condition is both unusual and easy to check for :)

We could do that but we'd have to make sure it applies to the last string written as well.

Reply:
> Is part of the proposal to drop this? I think plenty of people might still want it.

Currently, we are using five bytes of overhead per string on the string buffer. The null + a 32-bit counter. Dropping this to two bytes sounds good to me.

Reply:
> actually, as long as the next string is less than 0x100000000 bytes, if we don't add 32 to the "overflow length," all strings will be null terminated by default. And that condition is both unusual and easy to check for :)

Better: we can null terminate in the special code path that handles long strings. So, yeah, I thin we can realistically cut down the over per string from 5 bytes to 2 bytes.

Reply:
The new overhead per string is zero at the cost of doing something very complicated when the string is large.